main.go

```diff
@@ -2275,25 +2288,192 @@ The following options are recognized:
 				return err
 			}
 
-			s, err := ca.Secret(opt.SkipIfExists)
+			err = ca.SaveTo(v, opt.X509.Issue.SignedBy, opt.SkipIfExists)
+			if err != nil {
+				return err
+			}
+		}
+
+		err = cert.SaveTo(v, args[0], opt.SkipIfExists)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+
+	r.Dispatch("x509 reissue", &Help{
+		Summary: "Reissue X.509 Certificates and Certificate Authorities",
+		Usage:   "safe x509 reissue [OPTIONS] path/to/certificate",
+		Description: `
+Reissues an X.509 Certificate with a new key.
+
+The following options are recognized:
+
+  -b, --bits  N      RSA key strength, in bits.  The only valid
+                     arguments are 1024 (highly discouraged),
+                     2048 and 4096.  Defaults to 4096
+
+  -i, --signed-by   Path in the Vault where the CA certificate
+                    (and signing key) can be found.  If this is not
+                    provided, a sibling secret named 'ca' will used
+                    if it exists.
+
+  -t, --ttl         How long the new certificate will be valid
+                    for.  Specified in units h (hours), m (months)
+                    d (days) or y (years).  1m = 30d and 1y = 365d
+                    Defaults to the last TTL used to issue or renew
+                    the certificate.
+
+	`,
+	}, func(command string, args ...string) error {
+		rc.Apply(opt.UseTarget)
+
+		if len(args) != 1 {
+			r.ExitWithUsage("x509 reissue")
+		}
+
+		v := connect(true)
+
+		/* find the Certificate that we want to renew */
+		s, err := v.Read(args[0])
+		if err != nil {
+			return err
+		}
+		cert, err := s.X509()
+		if err != nil {
+			return err
+		}
+
+		/* find the CA */
+		ca, caPath, err := findSigningCa(v, cert, args[0], opt.X509.Reissue.SignedBy)
+		if err != nil {
+			return err
+		}
+
+		// Get new expiry date
+		var ttl time.Duration
+		if opt.X509.Reissue.TTL == "" {
+			ttl = cert.Certificate.NotAfter.Sub(cert.Certificate.NotBefore)
+		} else {
+			ttl, err = duration(opt.X509.Reissue.TTL)
 			if err != nil {
 				return err
 			}
-			err = v.Write(opt.X509.Issue.SignedBy, s)
+		}
+
+		// Get signing key bit length
+		var bits int
+		if opt.X509.Reissue.Bits != 0 {
+			bits = opt.X509.Reissue.Bits
+			if bits != 1024 && bits != 2048 && bits != 4096 {
+				return fmt.Errorf("Bits must be one of 1024, 2048 or 4096")
+			}
+		} else {
+			bits = cert.PrivateKey.N.BitLen()
+		}
+
+		// Generate new key with same bit length.
+		fmt.Printf("\nGenerating new %d-bit key...\n", bits)
+		newKey, err := rsa.GenerateKey(rand.Reader, bits)
+		if err != nil {
+			return err
+		}
+		cert.PrivateKey = newKey
+		err = ca.Sign(cert, ttl)
+		if err != nil {
+			return err
+		}
+		if caPath != args[0] {
+			err = ca.SaveTo(v, caPath, false)
 			if err != nil {
 				return err
 			}
 		}
 
-		s, err := cert.Secret(opt.SkipIfExists)
+		err = cert.SaveTo(v, args[0], false)
```

Comment:
This brings up an interesting point.

Should we complain if someone tries to `safe x509 reissue --skip-if-exists`?